src/services/jsTyping.ts

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft. All rights reserved. Licensed under the Apache License, Version 2.0. 
+// See LICENSE.txt in the project root for complete license information.
+
+/// <reference path='services.ts' />
+
+/* @internal */
+namespace ts.JsTyping {
+
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    var _host: HostType;
+    var _safeList: Map<string>;
+
+    // a typing name to typing file path mapping
+    var inferredTypings: Map<string> = {};
+
+    const notFoundTypingNames: string[] = [];
+
+    function tryParseJson(jsonPath: string): any {
+        if (_host.fileExists(jsonPath)) {
+            try {
+                // Strip out single-line comments
+                const contents = _host.readFile(jsonPath).replace(/^\/\/(.*)$/gm, "");
+                return JSON.parse(contents);
+            }
+            catch (e) { }
+        }
+        return undefined;
+    }
+
+    /**
+     * @param cachePath is the path to the cache location, which contains a tsd.json file and a typings folder
+     * @param fileNames are the file names that belongs to the same project
+     * @param globalCachePath is used 1. to get safe list file path 2. when we can't locate the project root path 
+     * @param projectRootPath
+     * @param typingOptions
+     * @param compilerOptions is used for typing inferring. 
+     */
+    export function discoverTypings(
+        host: HostType,
+        fileNames: string[],
+        globalCachePath: Path,
+        cachePath: Path,
+        typingOptions: TypingOptions,
+        compilerOptions?: CompilerOptions)
+        : { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[], newTsdJsonPath?: string } {
+
+        _host = host;
+        // Clear inferred typings map
+        inferredTypings = {};
+
+        cachePath = cachePath ? cachePath : globalCachePath;
+        // Only infer typings for .js and .jsx files

```

Comment:
Again, not sure we want to restrict this to only projects containing JavaScript files. If I have a project with TypeScript files, and I set `AutoAcquireTypings` to `true`, shouldn't it also get the typings for my Node modules as it would a project with JavaScript files?

We also don't know which files are using which libraries, so if I have a ton of code in TypeScript using libraries and modules, it won't get any typings automatically, but even if I add one JavaScript file (even if it's empty), we'll go get all the typings for your project?? That seems a little weird right?